extension/api.js

```diff
@@ -0,0 +1,193 @@
+/* global ExtensionCommon, ExtensionAPI, Services, XPCOMUtils */
+const { AddonManager } = ChromeUtils.import(
+  "resource://gre/modules/AddonManager.jsm"
+);
+const { WebRequest } = ChromeUtils.import(
+  "resource://gre/modules/WebRequest.jsm"
+);
+
+XPCOMUtils.defineLazyGlobalGetters(this, ["ChannelWrapper"]);
+
+XPCOMUtils.defineLazyServiceGetter(
+  this,
+  "eTLD",
+  "@mozilla.org/network/effective-tld-service;1",
+  "nsIEffectiveTLDService"
+);
+
+XPCOMUtils.defineLazyGetter(global, "searchInitialized", () => {
+  if (Services.search.isInitialized) {
+    return Promise.resolve();
+  }
+
+  return ExtensionUtils.promiseObserved(
+    "browser-search-service",
+    (_, data) => data === "init-complete"
+  );
+});
+
+const SEARCH_TOPIC_ENGINE_MODIFIED = "browser-search-engine-modified";
+
+this.addonsSearchExperiment = class extends ExtensionAPI {
+  getAPI(context) {
+    const { extension } = context;
+
+    return {
+      addonsSearchExperiment: {
+        // `getMatchPatterns()` returns a map where each key is an URL pattern
+        // to monitor and its corresponding value is a list of add-on IDs
+        // (search engines).
+        //
+        // Note: We don't return a simple list of URL patterns because the
+        // background script might want to lookup add-on IDs for a given URL.
+        async getMatchPatterns() {
+          const patterns = {};
+
+          try {
+            await searchInitialized;
+            const visibleEngines = await Services.search.getVisibleEngines();
```

Comment:
Per discussions, I think it is better to use `getEngines()` here. Might want to rename the variable as well.